### ISSUE TITLE
Increase default VTs from 6 to 12 to improve UX and also avoid throwing user into a blinking console

### ISSUE THREAD

**Issue by C0rn3j** [systemd/systemd#30923]
### systemd version the issue has been seen with

255.2

### Used distribution

Arch Linux

### Linux kernel version used

6.6.8-arch1-1

### CPU architectures issue was seen on

x86_64

### Component

systemd-logind

### Expected behaviour you didn't see

When attempting to switch to tty7-tty12 from X or a Wayland compositor via XKB, one ends up in a usable console.

### Unexpected behaviour you saw

Black screen with a blinking underscore in top left from tty7-tty12.

This confuses users ignorant of the complexities of VTs, I ran into this myself and see [other people](https://github.com/systemd/systemd/assets/1641362/54c36242-78c2-4685-91c4-4c60ca74346a) run into it from time to time.

It is possible to change this default on a given system, but by the point one is aware of this, they already ran into this issue.

### Steps to reproduce the problem

Press <kbd>CTRL+ALT+F7</kbd> on X or a Wayland compositor.

This does not happen when switching from a tty as the kernel has a function to check whether a switch makes sense, such functionality seems missing in userland so systemd defaults end up with bad UX.

---

Systemd by default allocates 6 VTs, [NAutoVTs/ReserveVT](https://www.freedesktop.org/software/systemd/man/latest/logind.conf.html), this default was chosen ["in order to be close to the traditional default configuration of Linux systems"](https://0pointer.de/blog/projects/serial-console.html).

The traditional systems usually bound to the first 6 VTs and binding more was probably undesirable, as doing so usually started all 6 getty instances immediately, consuming resources. 

Systemd nowadays starts up VTs dynamically - it does not start getty on them until the tty is accessed, so increasing this default seems okay from a resource-consumption point.  
*Not to mention that RAM was probably counted in MBs back when 6 gettys were being launched*

Here's a simplified diagram of switching from `tty` vs switching from `X / Wayland_compositor`:

![image](https://github.com/systemd/systemd/assets/1641362/40d7af99-beaf-430f-8918-71eb4e21142d)

*[WIP, lost the file - Version with clickable links](https://cloud.rys.pw/s/BFd4nJPtr7TkcME)*

As can be seen, from tty there is an additional check [vc_cons_allocated()](https://github.com/torvalds/linux/blob/052d534373b7ed33712a63d5e17b2b6cdbce84fd/drivers/tty/vt/keyboard.c#L557-L589), so the kernel's input processing will not switch to a console that's not allocated, whereas from X or a Wayland compositor the tty is changed into directly, irrelevant of whether or not it makes sense to do so.

There does not seem to be a way to check whether a console is allocated from userspace.   
Bumping the systemd default seems like the most sensible option to me, even if the F7-F12 correctly checked for allocations, logind would still be halving the available key combinations - keybinds are already there and hardcoded to tty switching.

Ideally, a new kernel ioctl should be added that provides the same information as `vc_cons_allocated()` does, and systemd-logind should change the switch check (seemingly [here](https://github.com/systemd/systemd/blob/b24db7052e1c660d01f030eff1efa8c708ba551b/src/login/logind-seat.c#L286-L304)) to it.

I am not aware of whether or not binding to more VTs could break anything, hopefully there isn't some legacy reason that I am overlooking which would make it impossible.

**Why 12 and not a different number? 24? 35? 63?**

- The kernel supports up to 63 consoles, everything in the VT subsystem and in kbd is wired up properly to support that, systemd-logind seems to also support all 63
- X (and presumably wayland compositors) supports F35 key as the highest one
- The default kernel keymap has mappings up to console 24
- USB HID has F24 key as the maximum
- libxkbcommon and all of its consumers by induction have hardcoded mappings, only up to console 12

There's no point in binding on anything higher as a user can't end up there by pressing a keybind, only when directly requested through something like `chvt`. Pressing <kbd>CTRL+ALT+F13</kbd> and up does nothing, thus to prevent this issue in default configuration, it's only relevant to bump this up to 12.

TL;DR three things should happen for perfection *(most likely with only the first one relevant for the scope of this issue)*
* Bump default VTs in systemd-logind to 12 to both avoid the bug and provide better UX, matching existing hardcoded XKB bindings
* Implement an ioctl in linux that provides the currently-internal `vs_cons_allocated()` check
* Implement the new ioctl in [systemd-logind](https://github.com/systemd/systemd/blob/b24db7052e1c660d01f030eff1efa8c708ba551b/src/login/logind-seat.c#L296)

---

Big thanks to @9ary, I wouldn't be able to figure out most of the information in this report without them.

**Comment by 9ary** [comment]
Technical precision: VT_ACTIVATE always tries to allocate the destination console. VT's own input logic does not, so the switch fails (silently, since nothing checks the return value of set_console in the kernel).

**Comment by poettering** [comment]
I am not sure I follow. First of all, in 2023 the console is not really maintained in the kernel, it's lagging behind in so many ways, hence I doubt people should use it more or anything about it could ever be user friendly.

Also, display managers that support multiple users to log in in parallel generally allocate multiple VTs for them from tty7 upwards. If we'd block them via gettys, then this would be very confusing as you couldn't switch to those sessions with a regular keyboard that easily anymore.

Hence, I am not convinced at all that this is desirable. If you like more text consoles, then change the setting, but I am very much sure the upstream default should stay as it is.

Or what am I not grokking here?

**Comment by C0rn3j** [comment]
> display managers that support multiple users to log in in parallel generally allocate multiple VTs for them from tty7 upwards

That I was not aware of, so the question here is if it's worth improving the UX a little bit while possible breaking DMs in niche(?) situations.  

Some existing DMs possibly aren't already checking if a tty is in use by either `VT_GETSTATE`(limited to tty1-tty15) and or `VT_OPENQRY` and thus could potentially conflict if a user starts a getty session on tty7+ before a DM is loaded.  
  
This would be an unusual scenario I suppose. User kills DM, switches to tty7 to start getty, tries to start DM... or DM possibly crashes and getty now fights with the DM for what launches first?

---

The issue with `logind` [happily switching to an unused and unclaimed empty tty](https://github.com/systemd/systemd/blob/b24db7052e1c660d01f030eff1efa8c708ba551b/src/login/logind-seat.c#L286-L304) is a thing that should be fixed either way.  
It possibly requires a new ioctl like `VT_GETALLOCATED`, as I am not aware of one that would be able to report tty allocation and not just usage.
 
If desired I can open a separate issue for it.

**Comment by 9ary** [comment]
> It possibly requires a new ioctl like `VT_GETALLOCATED`, as I am not aware of one that would be able to report tty allocation and not just usage.

Allocation probably doesn't actually matter. The kernel's behavior is not necessarily the best example to follow anyway. The console being allocated only means that it's been opened before.
In reality, we need to check whether the VT we're switching to is actually going to be useful. This is the case when autovt is configured for that VT, or when the console is open.

Something to this effect (untested):
```diff
diff --git a/src/login/logind-core.c b/src/login/logind-core.c
index 582cbcbea8..0b954443bb 100644
--- a/src/login/logind-core.c
+++ b/src/login/logind-core.c
@@ -494,7 +494,7 @@ int config_parse_n_autovts(
         return 0;
 }
 
-static int vt_is_busy(unsigned vtnr) {
+int vt_is_busy(unsigned vtnr) {
         struct vt_stat vt_stat;
         int r;
         _cleanup_close_ int fd = -EBADF;
diff --git a/src/login/logind-seat.c b/src/login/logind-seat.c
index 8d875d2ea7..e984e2623f 100644
--- a/src/login/logind-seat.c
+++ b/src/login/logind-seat.c
@@ -294,7 +294,8 @@ int seat_switch_to(Seat *s, unsigned num) {
         session = seat_get_position(s, num);
         if (!session) {
                 /* allow switching to unused VTs to trigger auto-activate */
-                if (seat_has_vts(s) && num < 64)
+                bool busy = num >= 16 || vt_is_busy(num);
+                if (seat_has_vts(s) && num < 64 && (num <= s->manager->n_autovts || busy))
                         return chvt(num);
 
                 return -EINVAL;
diff --git a/src/login/logind.h b/src/login/logind.h
index cac6a30357..5b8dc28ea6 100644
--- a/src/login/logind.h
+++ b/src/login/logind.h
@@ -188,3 +188,5 @@ int manager_setup_wall_message_timer(Manager *m);
 bool logind_wall_tty_filter(const char *tty, bool is_local, void *userdata);
 
 int manager_read_efi_boot_loader_entries(Manager *m);
+
+int vt_is_busy(unsigned vtnr);
```

**Comment by C0rn3j** [comment]
@9ary I have tested the patch (I know, took time) and it does work, I can no longer switch to tty7 (nothing runs there) from a Plasma or SDDM session.

@poettering Would a PR with the above patch make it in, or is there any issue with the code as is?
